File: samples-java/src/main/java/io/github/ilaborie/loops/samples/recursion/Filter.java
package io.github.ilaborie.loops.samples.recursion;

import io.github.ilaborie.loops.samples.Element;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static io.github.ilaborie.loops.samples.Elements.getSomeElements;

public class Filter {

    public static void main(String[] args) {
        List<Element> elements = getSomeElements();
        List<Element> result = filter(elements);
        result.forEach(System.out::println);
    }

public static List<Element> filter(List<Element> elements) {
    if (elements.isEmpty()) {
        return Collections.emptyList();
    }

    Element head = elements.get(0);
    List<Element> tail = elements.subList(1, elements.size());

    List<Element> tailFiltered = filter(tail);
    List<Element> result;
    if (head.getName().startsWith("a")) {
        result = new ArrayList<>();
        result.add(head);
        result.addAll(tailFiltered);
    } else {
        result = tailFiltered;
    }
    return result;
}

}

File: samples-kotlin/src/main/kotlin/io/github/ilaborie/loops/samples/recursion/filter.kt
package io.github.ilaborie.loops.samples.recursion

import io.github.ilaborie.loops.samples.ElementKt
import io.github.ilaborie.loops.samples.getSomeElementsKt


fun main() {
    val elements = getSomeElementsKt()
    val result = filter(elements)
    result.forEach { println(it) }
}

fun filter(elements: List<ElementKt>): List<ElementKt> {
    if (elements.isEmpty()) {
        return emptyList()
    }

    val head = elements.first()
    val tail = elements.drop(1)

    val filterTail = filter(tail)
    return if (head.name.startsWith("a")) {
        listOf(head) + filterTail
    } else {
        filterTail
    }
}

fun filterAlt(elements: List<ElementKt>): List<ElementKt> =
    when {
        elements.isEmpty()                    ->
            emptyList()
        elements.first().name.startsWith("a") ->
            listOf(elements.first()) + filterAlt(elements.drop(1))
        else                                  ->
            filterAlt(elements.drop(1))
    }
